// link-check/index.js
"use strict";

var stream = require('stream');
var request = require('request');

function LinkCheckResult(link, statusCode, err) {
    if (!(this instanceof LinkCheckResult)) {
        return new LinkCheckResult(link, statusCode, err);
    }

    this.link = link;
    this.statusCode = statusCode || 0;
    this.err = err || null;
    this.status = (this.statusCode === 200) ? 'alive' : 'dead';

    return this;
}

module.exports = function linkCheck(link, callback) {
    var options = {
        headers: {
            // override 'User-Agent' (some sites return `401` when the user-agent isn't a web browser)
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36',
			strictSSL: false //, proxy: 'http://proxy:8080'
        },
    };
    request.head(link, options, function (err, res, body) {
        if (!err && res.statusCode === 200) {
            callback(null, new LinkCheckResult(link, res ? res.statusCode : 0, err)); // alive, returned 200 OK
            return;
        }

        // if HEAD fails (405 Method Not Allowed, etc), try GET
        request.get(link, options, function (err, res) {
            callback(null, new LinkCheckResult(link, res ? res.statusCode : 0, err));
        }).pipe(new stream.Writable({ write: function (chunk, encoding, next) { next(); } }));
    });
};
